### DISCUSSION TITLE
Different compiler for build targets inside a subdir

### DISCUSSION THREAD

**Discussion by mahmnv** [mesonbuild/meson#12183] [Q&A] [answered]
Hi, is there any way to use a different compiler inside a subdir or for some selected build targets? Meson currently compiles all *.c files with same compiler. For an embedded application, I need some files to be compiled with a different compiler. Is this possible without separating code into a different project/subproject? Also, is there a way to 'compile' some C files and 'cross-compile' some other C files with a different compiler as part of same build process? That also will solve my usecase.

**Comment by xclaesse** [comment]
You can definitely have different native and cross compiler. Targets that should be compiled for build machine should have `native: true` keyword argument.

**Comment by eli-schwartz** [comment, marked as answer]
Meson enforces consistency at the "machine" level. Mixing and matching different compilers for the same language is forbidden at the generic level -- you can't build one program with amd64 GCC and the other with amd64 clang -- but specifically allowed as part of cross build support.

See https://mesonbuild.com/Cross-compilation.html

Cross compile builds need to be specifically set up with a toolchain (machine) file defining the cross-specific compiler. Build targets that need to be built for the build (native) machine as part of the same `ninja` invocation will instead use `$CC_FOR_BUILD` or whatever the default is on PATH.

**Reply by mahmnv** [reply]
> Meson enforces consistency at the "machine" level. Mixing and matching different compilers for the same language is forbidden at the generic level -- you can't build one program with amd64 GCC and the other with amd64 clang -- but specifically allowed as part of cross build support.

Thanks for the answer. I'm assuming the same applies for subprojects also. ie, main project with amd64 gcc and subproject with amd64 clang is not supported. Correct?

**Reply by eli-schwartz** [reply]
Yes, compilers are global in scope and each subproject is built with the same per-language (C vs C++ vs Fortran) and per-machine (native build vs cross host) compiler.

**Reply by mahmnv** [reply]
@eli-schwartz This is an issue for adopting meson. Do you see meson upstream accepting PR for a feature with which you can override language for a subdirectory like 

`subdir('dir_name', lang='xyz')`

This will be helpful if there are two languages in meson and both languages have same file extension. For eg, say there are two languages `A` and `B` with `.x` extension and there's a project which contains source files of both languages. Currently meson cannot do this as it'll select one single compiler for compiling all `.x` files.

**Reply by xclaesse** [reply]
> This will be helpful if there are two languages in meson and both languages have same file extension.

Something that has been proposed for that case is `files('foo.x', language: 'c')`. I don't think there is opposition against a PR doing that.

**Reply by eli-schwartz** [reply]
> This will be helpful if there are two languages in meson and both languages have same file extension. For eg, say there are two languages `A` and `B` with `.x` extension and there's a project which contains source files of both languages. Currently meson cannot do this as it'll select one single compiler for compiling all `.x` files.

It would be helpful for discussion to have real world examples. Off the top of my head I can't think of any, though (at least not in languages that meson currently knows about). Could you provide an example?

> Something that has been proposed for that case is `files('foo.x', language: 'c')`. I don't think there is opposition against a PR doing that.

My understanding was that it was proposed a couple times and rejected by e.g. @jpakkane....

**Reply by mahmnv** [reply]
Can I know why it was rejected before?
